src/actions/attack.cpp

```diff
@@ -143,7 +143,7 @@ battle_context_unit_stats::battle_context_unit_stats(const unit& u,
 	backstab_pos = is_attacker && backstab_check(u_loc, opp_loc, units, resources::gameboard->teams());
 	rounds = weapon->get_specials("berserk").highest("value", 1).first;
 	if(weapon->combat_ability("berserk", 1).second) {
-		rounds = weapon->combat_ability("berserk", 1).first;
+		rounds = std::max(1, weapon->combat_ability("berserk", 1).first);
```

Comment:
When can this happen?

Since you use combat_ability could you write documentation for it? What's the int and bool that it returns?

Reply:
Ah, I just saw that you wrote in the issue title that it can happen if the value in cfg file is <1. What does the engine do on such a cfg file? Does it crash or something? What's the problem, and why is this patch the correct fix?

Reply:
@jostephd if the berserk ability has a negative value, then the fight continues until the death of one of the combatants, while one should have only one round of combat. if the value is 0 or 1, it is as if the capacity did not exist but not when the value is negative, hence the repair, when the Boolean value, I was afraid that the calculation when the berserk capacity is absent does not affect a unit having a berserk attack, and as for drains, I conditioned the calculation to the presence of the ability

Reply:
> if the berserk ability has a negative value, then the fight continues until the death of one of the combatants, while one should have only one round of combat.

Why should the fight have only one round of combat when `value` is negative? https://wiki.wesnoth.org/AbilitiesWML#Extra_keys_used_by_the_.5Bberserk.5D_special does not document _any_ meaning for value<0. I think it's just an error in WML.

> if the value is 0 or 1, it is as if the capacity did not exist

Please explain this in more words. What does a `[berserk]` special with `value=0` do? What does a `[berserk]` special with `value=1` do? The former should effectively disable the attack and the latter should keep the attack enabled as a one round attack (but still cumulate with other `[berserk]` specials, if any).

> when the Boolean value, I was afraid that the calculation when the berserk capacity is absent does not affect a unit having a berserk attack, and as for drains, I conditioned the calculation to the presence of the ability

I wasn't criticizing your use of the boolean value. I was asking you to write documentation for the `combat_ability` function since it lacks documentation and you seem to know what it does.

Reply:
> I wasn't criticizing your use of the boolean value. I was asking you to write documentation for the `combat_ability` function since it lacks documentation and you seem to know what it does.

To be clear: documentation for the function itself, not for the code that calls it.

Reply:
when i put value=-1 in berserk abilities, the fight continue until the death of a combattant, and round>1 then what value<1, if someone put that case then the effect will be different what expected.
for the documentation, how must make that?

Reply:
if value=0 do not disable fight in special both ability, the same problems showed when special berserk value<0, the fight continue until death of a combattant, like if it was not limit value, and it is why the negative value in berserk must be ignored, in special or ability. i will make a nuew PR for repair that .

Reply:
> the fight continue until the death of a combattant

That would be because `battle_context_unit_stats::rounds` is an `unsigned`, so assigning -1 to it sets it to the largest representable integer. That means the battle will have a finite but very large number of rounds. In practice, the loop should terminate soon enough by the death of one combatant (unless they both have 0% cth, or there's some sort of [drains] or [heal_on_hit] special, or something).

> and round>1 then what value<1, if someone put that case then the effect will be different what expected.

What was expected? value=-1 could mean either "no limit on the number of rounds", "limit of zero rounds", "only valid with `cumulative=yes`", or "unspecified/undefined behavior". All of these positions are defensible.

**Question to developers**: What should `[berserk] value=-1 [/berserk]` do?

>  for the documentation, how must make that?

Add doxygen comments (start with `///`) to the declaration or definition of the function.

Reply:
> if value=0 do not disable fight in special both ability, the same problems showed when special berserk value<0, the fight continue until death of a combattant, like if it was not limit value, and it is why the negative value in berserk must be ignored, in special or ability. i will make a nuew PR for repair that .

Sorry, I don't understand. What is the **current** behavior with value=0? What do you think the behavior should be?

Reply:
value=-1 meaning infinite rounds would be my expectation.